src/Riok.Mapperly/MapperGenerator.cs

```diff
@@ -22,39 +22,63 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
         var mapperClassDeclarations = SyntaxProvider.GetClassDeclarations(context);
 
         var compilationAndMappers = context.CompilationProvider.Combine(mapperClassDeclarations.Collect());
-        context.RegisterImplementationSourceOutput(compilationAndMappers, static (spc, source) => Execute(source.Left, source.Right, spc));
+        var mappersWithErrors = compilationAndMappers.Select((x, token) => BuildDescriptors(x.Left, x.Right, token));
+
+        // output the diagnostics
+        context.ReportDiagnostics(mappersWithErrors.Select(static (source, _) => source.Diagnostics));
+
+        // split into mapper name pairs
+        var mappers = mappersWithErrors.SelectMany((x, _) => x.Mappers);
+
+        context.RegisterImplementationSourceOutput(
+            mappers,
+            static (spc, source) =>
+            {
+                spc.AddSource(source.FileName, SourceText.From(source.Body.NormalizeWhitespace().ToFullString(), Encoding.UTF8));
```

Comment:
A lot going on on this line, probably split it into multiple lines to increase readability.